### ISSUE TITLE
Keras not using GPU AWS EC2 g3s.xlarge w/Tesla GPU

### ISSUE THREAD

**Issue by eafpres** [rstudio/keras#817]
I have an AWS EC2 Linux instance (g3s.xlarge) that has a Tesla M60 GPU.  I am running RStudio on it (this is a Louis Aslett AMI).
I ran
`install.packages("keras")`  
`install_keras(tensorflow = "gpu")`
with no errors
When I tried to run my code, I got the tensorflow not found error.
I then ran
`library(tensorflow)`
`install_tensorflow(method = "virtualenv")`
and the code runs.
However, it is not using the GPU

I checked the hardware and the GPU is there, and below I show the keras/tensorflow init messages and the messages when training starts.

sudo lshw
        *-display:1 UNCLAIMED
             description: VGA compatible controller
             **product: GM204GL [Tesla M60]
             vendor: NVIDIA Corporation**
             physical id: 1e
             bus info: pci@0000:00:1e.0
             version: a1
             width: 64 bits
             clock: 33MHz
             capabilities: pm msi pciexpress vga_controller bus_master cap_list
             configuration: latency=0
             resources: memory:94000000-94ffffff memory:80000000-8fffffff memory:92000000-93ffffff ioport:c100(size=128)

>   set.seed(20190614)
>   use_session_with_seed(20190614, disable_gpu = FALSE, disable_parallel_cpu = FALSE)

2019-06-18 02:11:18.629892: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-06-18 02:11:18.634559: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2300025000 Hz
2019-06-18 02:11:18.634863: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x129d4310 executing computations on platform Host. Devices:
2019-06-18 02:11:18.634882: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
Set session seed to 20190614

WARNING: The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
If you depend on functionality not listed there, please file an issue.

Train on 800 samples, validate on 200 samples
Epoch 1/500
800/800 [==============================] - 0s 468us/sample - loss: 0.0137 - mean_squared_error: 0.0137 - val_loss: 0.0992 - val_mean_squared_error: 0.0992

**Comment by dfalbel** [comment]
If it didn't work with `install_tensorflow(method = "virtualenv", version = "gpu")` it probably means that you don't have the right NVIDIA drivers installed in that machine. 

Can you tell me the results of rrunning `nvidia-smi` in your terminal?

**Comment by eafpres** [comment]
That definitely looks like a problem; now need to figure out how to update this instance.

NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.

**Comment by eafpres** [comment]
Updated drivers with sudo install nvidia-375
appeared to work
rebooted instance
removed keras & tensorflow
`remove.packages("keras")`
`remove.packages("tensorflow")`
reinstalled with
`install.packages("keras")`
`install_keras(tensorflow = "gpu")`
no errors or warnings
tried to run my code
`Error: Installation of TensorFlow not found.

Python environments searched for 'tensorflow' package:
 /home/rstudio/.virtualenvs/r-tensorflow/bin/python
 /usr/bin/python2.7
 /usr/bin/python3.5

You can install TensorFlow using the install_tensorflow() function.`

so I ran:
`library(tensorflow)`
`install_tensorflow(version = "gpu")`

`Creating virtualenv for TensorFlow at  ~/.virtualenvs/r-tensorflow 
New python executable in /home/rstudio/.virtualenvs/r-tensorflow/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
Running virtualenv with interpreter /usr/bin/python
Installing TensorFlow ...
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Collecting tensorflow-gpu==1.13.1
  Using cached https://files.pythonhosted.org/packages/24/8c/459eb5fdaa8666010b6a32244ddb92f4015cfc54156d612959625096601d/tensorflow_gpu-1.13.1-cp27-cp27mu-manylinux1_x86_64.whl
Collecting h5py
  Using cached https://files.pythonhosted.org/packages/53/08/27e4e9a369321862ffdce80ff1770553e9daec65d98befb2e14e7478b698/h5py-2.9.0-cp27-cp27mu-manylinux1_x86_64.whl
Collecting pyyaml
Collecting requests
  Using cached https://files.pythonhosted.org/packages/51/bd/23c926cd341ea6b7dd0b2a00aba99ae0f828be89d72b2190f27c11d4b7fb/requests-2.22.0-py2.py3-none-any.whl
Collecting Pillow
  Using cached https://files.pythonhosted.org/packages/b6/4b/5adc1109908266554fb978154c797c7d71aba43dd15508d8c1565648f6bc/Pillow-6.0.0-cp27-cp27mu-manylinux1_x86_64.whl
Collecting keras
  Using cached https://files.pythonhosted.org/packages/5e/10/aa32dad071ce52b5502266b5c659451cfd6ffcbf14e6c8c4f16c0ff5aaab/Keras-2.2.4-py2.py3-none-any.whl
Collecting tensorflow-hub
  Using cached https://files.pythonhosted.org/packages/10/5c/6f3698513cf1cd730a5ea66aec665d213adf9de59b34f362f270e0bd126f/tensorflow_hub-0.4.0-py2.py3-none-any.whl
Collecting tensorflow-probability
  Using cached https://files.pythonhosted.org/packages/4e/93/191e42ca27786d6875f74e7b756e34ef42f385f6d250bfc28aa48a1d1072/tensorflow_probability-0.6.0-py2.py3-none-any.whl
Collecting scipy
  Using cached https://files.pythonhosted.org/packages/1d/f6/7c16d60aeb3694e5611976cb4f1eaf1c6b7f1e7c55771d691013405a02ea/scipy-1.2.2-cp27-cp27mu-manylinux1_x86_64.whl
Collecting astor>=0.6.0 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/d1/4f/950dfae467b384fc96bc6469de25d832534f6b4441033c39f914efd13418/astor-0.8.0-py2.py3-none-any.whl
Collecting enum34>=1.1.6 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/c5/db/e56e6b4bbac7c4a06de1c50de6fe1ef3810018ae11732a50f15f62c7d050/enum34-1.1.6-py2-none-any.whl
Collecting protobuf>=3.6.1 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/b2/a8/ad407cd2a56a052d92f602e164a9e16bede22079252af0db3838f375b6a8/protobuf-3.8.0-cp27-cp27mu-manylinux1_x86_64.whl
Collecting keras-preprocessing>=1.0.5 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/28/6a/8c1f62c37212d9fc441a7e26736df51ce6f0e38455816445471f10da4f0a/Keras_Preprocessing-1.1.0-py2.py3-none-any.whl
Collecting gast>=0.2.0 (from tensorflow-gpu==1.13.1)
Collecting six>=1.10.0 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Collecting tensorboard<1.14.0,>=1.13.0 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/89/ac/48dd71c2bdc8d31e367f9b72f25ccb3b89bc6b9d664fee21f9a8efa5714d/tensorboard-1.13.1-py2-none-any.whl
Collecting wheel (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/bb/10/44230dd6bf3563b8f227dbf344c908d412ad2ff48066476672f3a72e174e/wheel-0.33.4-py2.py3-none-any.whl
Collecting absl-py>=0.1.6 (from tensorflow-gpu==1.13.1)
Collecting backports.weakref>=1.0rc1 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/88/ec/f598b633c3d5ffe267aaada57d961c94fdfa183c5c3ebda2b6d151943db6/backports.weakref-1.0.post1-py2.py3-none-any.whl
Collecting termcolor>=1.1.0 (from tensorflow-gpu==1.13.1)
Collecting numpy>=1.13.3 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/1f/c7/198496417c9c2f6226616cff7dedf2115a4f4d0276613bab842ec8ac1e23/numpy-1.16.4-cp27-cp27mu-manylinux1_x86_64.whl
Collecting tensorflow-estimator<1.14.0rc0,>=1.13.0 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/bb/48/13f49fc3fa0fdf916aa1419013bb8f2ad09674c275b4046d5ee669a46873/tensorflow_estimator-1.13.0-py2.py3-none-any.whl
Collecting grpcio>=1.8.6 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/8e/62/448199c50e05526f7fd336d8a87bd83a0c7d92188ae51ebea23aa5f3c37e/grpcio-1.21.1-cp27-cp27mu-manylinux1_x86_64.whl
Collecting mock>=2.0.0 (from tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/05/d2/f94e68be6b17f46d2c353564da56e6fb89ef09faeeff3313a046cb810ca9/mock-3.0.5-py2.py3-none-any.whl
Collecting keras-applications>=1.0.6 (from tensorflow-gpu==1.13.1)
Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 (from requests)
  Using cached https://files.pythonhosted.org/packages/e6/60/247f23a7121ae632d62811ba7f273d0e58972d75e58a94d329d51550a47d/urllib3-1.25.3-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests)
  Using cached https://files.pythonhosted.org/packages/69/1b/b853c7a9d4f6a6d00749e94eb6f3a041e342a885b87340b79c1ef73e3a78/certifi-2019.6.16-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2 (from requests)
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Collecting idna<2.9,>=2.5 (from requests)
  Using cached https://files.pythonhosted.org/packages/14/2c/cd551d81dbe15200be1cf41cd03869a46fe7226e7450af7a6545bfc474c9/idna-2.8-py2.py3-none-any.whl
Collecting setuptools (from protobuf>=3.6.1->tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/ec/51/f45cea425fd5cb0b0380f5b0f048ebc1da5b417e48d304838c02d6288a1e/setuptools-41.0.1-py2.py3-none-any.whl
Collecting futures>=3.1.1; python_version < "3" (from tensorboard<1.14.0,>=1.13.0->tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/2d/99/b2c4e9d5a30f6471e410a146232b4118e697fa3ffc06d6a65efde84debd0/futures-3.2.0-py2-none-any.whl
Collecting markdown>=2.6.8 (from tensorboard<1.14.0,>=1.13.0->tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/c0/4e/fd492e91abdc2d2fcb70ef453064d980688762079397f779758e055f6575/Markdown-3.1.1-py2.py3-none-any.whl
Collecting werkzeug>=0.11.15 (from tensorboard<1.14.0,>=1.13.0->tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/9f/57/92a497e38161ce40606c27a86759c6b92dd34fcdb33f64171ec559257c02/Werkzeug-0.15.4-py2.py3-none-any.whl
Collecting funcsigs>=1; python_version < "3.3" (from mock>=2.0.0->tensorflow-gpu==1.13.1)
  Using cached https://files.pythonhosted.org/packages/69/cb/f5be453359271714c01b9bd06126eaf2e368f1fddfff30818754b5ac2328/funcsigs-1.0.2-py2.py3-none-any.whl
Installing collected packages: astor, enum34, setuptools, six, protobuf, numpy, keras-preprocessing, gast, futures, markdown, grpcio, wheel, werkzeug, absl-py, tensorboard, backports.weakref, termcolor, funcsigs, mock, tensorflow-estimator, h5py, keras-applications, tensorflow-gpu, pyyaml, urllib3, certifi, chardet, idna, requests, Pillow, scipy, keras, tensorflow-hub, tensorflow-probability
Successfully installed Pillow-6.0.0 absl-py-0.7.1 astor-0.8.0 backports.weakref-1.0.post1 certifi-2019.6.16 chardet-3.0.4 enum34-1.1.6 funcsigs-1.0.2 futures-3.2.0 gast-0.2.2 grpcio-1.21.1 h5py-2.9.0 idna-2.8 keras-2.2.4 keras-applications-1.0.8 keras-preprocessing-1.1.0 markdown-3.1.1 mock-3.0.5 numpy-1.16.4 protobuf-3.8.0 pyyaml-5.1.1 requests-2.22.0 scipy-1.2.2 setuptools-41.0.1 six-1.12.0 tensorboard-1.13.1 tensorflow-estimator-1.13.0 tensorflow-gpu-1.13.1 tensorflow-hub-0.4.0 tensorflow-probability-0.6.0 termcolor-1.1.0 urllib3-1.25.3 werkzeug-0.15.4 wheel-0.33.4

Installation complete.


Restarting R session...`

tried to run my code, got this

`>   set.seed(20190614)
>   use_session_with_seed(20190614, disable_gpu = FALSE, disable_parallel_cpu = FALSE)
Error: Installation of TensorFlow not found.

Python environments searched for 'tensorflow' package:
 /home/rstudio/.virtualenvs/r-tensorflow/bin/python
 /usr/bin/python2.7
 /usr/bin/python3.5

You can install TensorFlow using the install_tensorflow() function.`

so on whim, I added

`library(tensorflow)`

to my code, reran, got this:

`Error: Python module tensorflow.python.keras was not found.

Detected Python configuration:

python:         /home/rstudio/.virtualenvs/r-tensorflow/bin/python
libpython:      /usr/lib/python2.7/config-x86_64-linux-gnu/libpython2.7.so
pythonhome:     /usr:/usr
virtualenv:     /home/rstudio/.virtualenvs/r-tensorflow/bin/activate_this.py
version:        2.7.12 (default, Nov 12 2018, 14:36:49)  [GCC 5.4.0 20160609]
numpy:          /home/rstudio/.virtualenvs/r-tensorflow/local/lib/python2.7/site-packages/numpy
numpy_version:  1.16.4
tensorflow:     /home/rstudio/.virtualenvs/r-tensorflow/local/lib/python2.7/site-packages/tensorflow

python versions found: 
 /home/rstudio/.virtualenvs/r-tensorflow/bin/python
 /usr/bin/python
 /usr/bin/python3`

**Comment by eafpres** [comment]
FYI, checking the driver after the reboot gives:

`Tue Jun 18 21:11:58 2019
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.130                Driver Version: 384.130                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla M60           Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   40C    P0    38W / 150W |      0MiB /  7613MiB |     94%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
rstudio@ip-172-31-23-238:/$`

**Comment by dfalbel** [comment]
The drivers seems to be installed now, but you also need to have CUDA installed. Here's the output of my nvidia-smi for you to compare:

![image](https://user-images.githubusercontent.com/4706822/59752635-ea8ca380-9282-11e9-8028-058963762609.png)

**Comment by eafpres** [comment]
After trying various options to no avail, I replaced this with a newer AMI that has CUDA-10.0 and cuDNN-7.5.1, and everything installed and worked first time.  I suspect that there may be an issue trying to use the Tesla M60 with CUDA-9.0 or something along those lines.
